File: plugins/fr.cea.nabla.javalib/src/fr/cea/nabla/javalib/linearalgebra/Vector.java
/*******************************************************************************
 * Copyright (c) 2020 CEA
 * This program and the accompanying materials are made available under the 
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * SPDX-License-Identifier: EPL-2.0
 * Contributors: see AUTHORS file
 *******************************************************************************/
package fr.cea.nabla.javalib.linearalgebra;

import java.util.Arrays;

import org.apache.commons.math3.linear.ArrayRealVector;
import org.apache.commons.math3.linear.OpenMapRealVector;

public class Vector 
{
	private Object lock = new Object();
	private org.apache.commons.math3.linear.RealVector nativeVector;

	public final int length;

	public Vector(org.apache.commons.math3.linear.RealVector nativeVector)
	{
		this.nativeVector = nativeVector;
		this.length = nativeVector.getDimension();
	}

	public static Vector createDenseVector(int size)
	{
		return new Vector(new ArrayRealVector(size));
	}

	public static Vector createSparseVector(int size)
	{
		return new Vector(new OpenMapRealVector(size));
	}

	public org.apache.commons.math3.linear.RealVector getNativeVector()
	{
		return nativeVector;
	}

	public double get(int i)
	{
		return nativeVector.getEntry(i);
	}

	public void set(int i, double value)
	{
		synchronized(lock) { nativeVector.setEntry(i, value); }
	}

	public void add(int i, double increment) 
	{
		synchronized(lock) { nativeVector.addToEntry(i, increment); }
	}

	@Override
	public String toString()
	{
		return Arrays.toString(nativeVector.toArray());
	}
}

File: plugins/nablacpp/nablalib/linearalgebra/stl/Vector.cc
/*******************************************************************************
 * Copyright (c) 2020 CEA
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * SPDX-License-Identifier: EPL-2.0
 * Contributors: see AUTHORS file
 *******************************************************************************/
#include "nablalib/linearalgebra/stl/Vector.h"

namespace nablalib::linearalgebra::stl
{

Vector::
Vector(const std::string& name, const int size)
: m_data(size) {}


Vector::
Vector(VectorType& v)
: m_data(v) {}


Vector::
~Vector() {}


Vector& Vector::
operator=(const Vector& val)
{
	m_data = val.m_data;
	return *this;
}


const int Vector::
getSize() const
{
  return m_data.size();
}


double Vector::
getValue(const int i) const
{
  return m_data[i];
}


void Vector::
setValue(const int i, double value)
{
	m_data[i] = value;
}

}
